hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java

```diff
@@ -317,30 +335,61 @@ private synchronized void addPendingData(Data data) {
         pendingData.add(data);
     }
 
+    private synchronized void addPendingEvent(Collection<Event> event) {
+        pendingEvent.addAll(event);
+    }
+
+    private synchronized void addPendingEvent(Event event) {
+        pendingEvent.add(event);
+    }
+
     private synchronized Collection<Data> getAndClearPendingData() {
         Collection<Data> result = new ArrayList<>(pendingData);
         pendingData.clear();
         return result;
     }
 
+    private synchronized Collection<Event> getAndClearPendingEvent() {
+        Collection<Event> result = new ArrayList<>(pendingEvent);
+        pendingEvent.clear();
+        return result;
+    }
+
     private class RulesInvoker extends TimerTask {
         @Override
         public void run() {
             int numTimeouts = checkPendingTimeouts();
 
-            if (!pendingData.isEmpty() || numTimeouts > 0) {
+            if (!pendingData.isEmpty() || !pendingEvent.isEmpty() || numTimeouts > 0) {
                 Collection<Data> newData = getAndClearPendingData();
+                Collection<Event> newEvent = getAndClearPendingEvent();
+
+                if (!newEvent.isEmpty()) {
+                    /*
+                        Incoming Events are added to the result events list.
+                        Events are all persisted at the end of the inference.
+                        With this we will have on events list:
+                        - Input events.
+                        - New events generates as result of the inference process.
+                     */
+                    events.addAll(newEvent);
+                }
```

Comment:
Original logic was:
- Events that are sent to the engine are persisted by the engine with the emitted one at the end of the inference.
- Events that are not sent to the engine are directly persisted.

Your comment could be to persist first and only let the engine to persist new events.
It could be valid, I don't think it change the algorithm, only the place where it is executed, so I'm +1 with the change if it is more clear.